README
Suriproxy, Copyright 2004 by Yves Junqueira
nictuku@users.sourceforge.net

#warning#
Suriproxy was made with my needs in sight, so you will probably be 
bothered with its lack of features and almost non existent 
documentation. In case I get some feedback, I could try and improve it.
########

Suriproxy is a fast pre-queue filter for Postfix that will inspect all
messages content and REJECT those that contains Spamvertised URLs. 
It addresses the false positives problem by REJECTing messages instead 
of discarding, or worse, bouncing them.

See www.surbl.org for SURBL servers to use.

In case of a Suriproxy failure during scan, Postfix will issue smtp error 
code to the client, asking it to try again later. Also, the postmaster 
will receive a mail regarding the error.
As all GPL software, Suriproxy comes with no warranty against damage
and data loss. I use it in my production servers, so I guess it will
not harm your system either.

Remote clients will keep trying to deliver the messages, so you
hopefully will not lose mail if Suriproxy fails.
As it is in early stages of development and tests, keep an
eye on the logs.

### known problems ###

If you use amavisd-new as your 'talk server', you may notice these
errors:

Oct  6 12:56:46 host amavis[23867]: (23867-05) ESMTP TROUBLE, ABORTING: 
timed out
Oct  6 12:56:46 host amavis[23867]: (23867-05) ESMTP: NOTICE: ABORTING 
the session: timed out

They happen after a denial if the client quits the connection. 
Suriproxy will not issue a QUIT to amavisd-new, instead just close the 
connection. Ignore it :)

### How to use it ###

First, change master.cf in case you use postfix:

 smtp      inet  n       -       n       -       -       smtpd
   -o smtpd_proxy_filter=127.0.0.1:10023

If you are using a "content_filter", like amavisd-new, you should 
replace it.

You might want to increase smtpd_proxy_timeout to a bigger value, in
main.cf:

smtpd_proxy_timeout = 60m

In my own experience, this timeout also applies to the smtp dialog between
postfix and the client. So, big attachments that take more than
$smtpd_proxy_timeout to arrive would trigger errors. It's a bandwidth
limit issue. It doesn't seem to be Suriproxy's performance fault.

Then, decide which will be your after-pre-queue daemon. I am using
amavisd-new. You could use a special postfix smtp server instance.

If your after-pre-queue daemon is running at tcp:10024, you will
start Suriproxy with

# /usr/local/sbin/suriproxy 127.0.0.1:10023 127.0.0.1:10024

It will stay in the foreground by default.
You should run it in the foreground some time, so you could
get any warnings and errors in STDERR.

You could use Suriproxy as a post-queue filter, but it still does not
feature mail quarantine or non-discarding options. So you would 
definetly LOSE MAIL. If you are aware of the risk, change $mode to a 
non-"transparent" value, and use content_filter instead of 
smtpd_proxy_filter in your master.cf.

Accepted recipient whitelist formats:

abuse@
domain.tld
.domain.tld (whitelists subdomains too)
abuse@domain.tld
abuse@.domain.tld

See Postfix' SMTPD_PROXY_README.


TODO:
 - Textify log messages errors.
 - Quit the connection cleanly with the talk-server.
 - Make a decent documentation and website

NOT TO DO:
 - Improve recipient whitelist so the same message can be delivered
   to foo but not to bar. 


ChangeLog
1.0:
Tue Nov 30 11:27:41 BRST 2004
- Better error treating
- Added support for "cloaked" URI queries using MD5.
  That way a provider can give the domain zones to his/her clients,
  but they will all be one-way "crypted" in MD5 hash, thus
  avoiding access to the proper list, which could be a security
  problem.
- Much better URI extracting code, using SpamAssassin's perl lib.

0.10:
Sun Oct 10 17:38:11 BRT 2004
- Recipient whitelist added.
- Bugfix of bugfix: server should really close the connection.
- Fixed a do_log bug that was screwing $msg's with %'s.

0.9:
Wed Oct  6 13:13:03 BRT 2004

- Added URI domain whitelist capability.
- Will try do decode MIME quoted-printed (not Base64 yet).
- Read messages in slurp mode.
- Better URI matching code using TLD's.
- Bugfix: Server was closing the connection after a denial.

0.8:
Thu Sep 30 13:24:38 BRT 2004

- Initial release


MSDW/SMTP/Server.pm
#   This code is Copyright (C) 2001 Morgan Stanley Dean Witter, and
#   is distributed according to the terms of the GNU Public License
#   as found at <URL:http://www.fsf.org/copyleft/gpl.html>.
#
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
# Written by Bennett Todd <bet@rahul.net>

package MSDW::SMTP::Server;
use IO::Socket;
use IO::File;

=head1 NAME

  MSDW::SMTP::Server --- SMTP server for content-scanning proxy

=head1 SYNOPSIS

  use MSDW::SMTP::Server;

  my $server = MSDW::SMTP::Server->new(interface => $interface,
				       port => $port);
  while (1) {
    # prefork here
    $server->accept([options]);
    # per-connect fork here
    $server->ok("220 howdy");
    while (my $what = $server->chat) {
      if ($what =~ /^mail/i) {
	if (isgood($server->{from})) {
	  $server->ok([ ack msg ]);
	} else {
	  $server->fail([ fail msg ]);
	}
      } elsif ($what =~ /^rcpt/i) {
	if (isgood(@{$server}{qw(from to)})) {
	  $sever->ok([ ack msg ]);
	} else {
	  $server->fail([ fail msg ]);
	}
      } elsif ($what =~ /^data/i) {
	if (isgood(@{$server}{qw(from to)})) {
	  # NB to is now an array of all recipients
	  $self->ok("354 natter on.");
	} else {
	  $self->fail;
	}
      } elsif ($what eq '.') {
        if (isgood(@server->{from,to,data})) {
	  $server->ok;
	} else {
	  $server->fail;
	}
      } else {
        # deal with other msg types as you will
	die "can't happen";
      }
      # process $server->{from,to,data} here
      $server->ok; # or $server->fail;
    }
  }

=head1 DESCRIPTION

MSDW::SMTP::Server fills a gap in the available range of Perl SMTP
servers. The existing candidates are not suitable for a
high-performance, content-scanning robust SMTP proxy. They insist on
heavy-weight structuring and parsing of the body, and they
acknowledge receipt of the data before returning control to the
caller.

This server simply gathers the SMTP acquired information (envelope
sender and recipient, and data) into unparsed memory buffers (or a
file for the data), and returns control to the caller to explicitly
acknowlege each command or request. Since acknowlegement or failure
are driven explicitly from the caller, this module can be used to
create a robust SMTP content scanning proxy, transparent or not as
desired.

=head1 METHODS

=over 8

=item new(interface => $interface, port => $port);

The interface and port to listen on must be specified. The interface
must be a valid numeric IP address (0.0.0.0 to listen on all
interfaces, as usual); the port must be numeric. If this call
succeeds, it returns a server structure with an open
IO::Socket::INET in it, ready to listen on. If it fails it dies, so
if you want anything other than an exit with an explanatory error
message, wrap the constructor call in an eval block and pull the
error out of $@ as usual. This is also the case for all other
methods; they succeed or they die.

=item accept([debug => FD]);

accept takes optional args and returns nothing. If an error occurs
it dies, otherwise it returns when a client connects to this server.
This is factored out as a separate entry point to allow preforking
(e.g. Apache-style) or fork-per-client strategies to be implemented
on the common protocol core. If a filehandle is passed for debugging
it will receive a complete trace of the entire SMTP dialogue, data
and all. Note that nothing in this module sends anything to the
client, including the initial login banner; all such backtalk must
come from the calling program.

=item chat;

The chat method carries the SMTP dialogue up to the point where any
acknowlegement must be made. If chat returns true, then its return
value is the previous SMTP command. If the return value begins with
'mail' (case insensitive), then the attribute 'from' has been filled
in, and may be checked; if the return value begins with 'rcpt' then
both from and to have been been filled in with scalars, and should
be checked, then either 'ok' or 'fail' should be called to accept
or reject the given sender/recipient pair. If the return value is
'data', then the attributes from and to are populated; in this case,
the 'to' attribute is a reference to an anonymous array containing
all the recipients for this data. If the return value is '.', then
the 'data' attribute (which may be pre-populated in the "new" or
"accept" methods if desired) is a reference to a filehandle; if it's
created automatically by this module it will point to an unlinked
tmp file in /tmp. If chat returns false, the SMTP dialogue has been
completed and the socket closed; this server is ready to exit or to
accept again, as appropriate for the server style.

The return value from chat is also remembered inside the server
structure in the "state" attribute.

=item ok([message]);

Approves of the data given to date, either the recipient or the
data, in the context of the sender [and, for data, recipients]
already given and available as attributes. If a message is given, it
will be sent instead of the internal default.

=item fail([message]);

Rejects the current info; if processing from, rejects the sender; if
processing 'to', rejects the current recipient; if processing data,
rejects the entire message. If a message is specified it means the
exact same thing as "ok" --- simply send that message to the sender.

=back

=cut

sub new {
    my ($this, @opts) = @_;
    my $class = ref($this) || $this;
    my $self = bless { @opts }, $class;
    $self->{sock} = IO::Socket::INET->new(
	LocalAddr => $self->{interface},
	LocalPort => $self->{port},
	Proto => 'tcp',
	Type => SOCK_STREAM,
	Listen => 65536,
	Reuse => 1,
    );
    die "$0: socket bind failure: $!\n" unless defined $self->{sock};
    $self->{state} = 'just bound',
    return $self;
}

sub accept {
    my ($self, @opts) = @_;
    %$self = (%$self, @opts);
    ($self->{"s"}, $self->{peeraddr}) = $self->{sock}->accept or
	die "$0: accept failure: $!\n";
    $self->{state} = ' accepted';
}


sub chat {
    my ($self) = @_;
    local(*_);
    if ($self->{state} !~ /^data/i) {
	return 0 unless defined($_ = $self->getline);
	s/[\r\n]*$//;
	$self->{state} = $_;
	if (s/^helo\s+//i || s/^ehlo\s+//i) {
	    s/\s*$//;s/\s+/ /g;
	    $self->{helo} = $_;
	} elsif (s/^xforward\s*//i) {
	    my (@attr) = split(/\s+/, $_);
	    my ($attr);
	    foreach $attr (@attr) {
		my ($name, $value) = ($attr =~ /^(.*?)=(.*)$/);
		$self->{xforward}->{lc($name)} = $value;
	    }
	} elsif (s/^rset\s*//i) {
	    delete $self->{to};
	    delete $self->{data};
	    delete $self->{recipients};
	    delete $self->{xforward};
	} elsif (s/^mail\s+from:\s*//i) {
	    delete $self->{to};
	    delete $self->{data};
	    delete $self->{recipients};
	    s/\s*$//;
	    $self->{from} = $_;
	} elsif (s/^rcpt\s+to:\s*//i) {
	    s/\s*$//; s/\s+/ /g;
	    $self->{to} = $_;
	    push @{$self->{recipients}}, $_;
	} elsif (/^data/i) {
	    $self->{to} = $self->{recipients};
	}
    } else {
	if (defined($self->{data})) {
	    $self->{data}->seek(0, 0);
	    $self->{data}->truncate(0);
	} else {
	    $self->{data} = IO::File->new_tmpfile;
	}
	while (defined($_ = $self->getline)) {
	    if ($_ eq ".\r\n") {
		$self->{data}->seek(0,0);
		return $self->{state} = '.';
	    }
	    s/^\.\./\./;
	    $self->{data}->print($_) or die "$0: write error saving data\n";
	}
	return(0);
    }
    return $self->{state};
}

sub getline {
    my ($self) = @_;
    local ($/) = "\r\n";
    return $self->{"s"}->getline unless defined $self->{debug};
    my $tmp = $self->{"s"}->getline;
    $self->{debug}->print($tmp) if ($tmp);
    return $tmp;
}

sub print {
    my ($self, @msg) = @_;
    $self->{debug}->print(@msg) if defined $self->{debug};
    $self->{"s"}->print(@msg);
}

sub ok {
    my ($self, @msg) = @_;
    @msg = ("250 ok.") unless @msg;
    $self->print("@msg\r\n") or
	die "$0: write error acknowledging $self->{state}: $!\n";
}

sub fail {
    my ($self, @msg) = @_;
    @msg = ("550 no.") unless @msg;
    $self->print("@msg\r\n") or
	die "$0: write error acknowledging $self->{state}: $!\n";
}

1;


suriproxy.pl
#!/usr/bin/perl -W

# Suriproxy Copyright 2004, Yves Junqueira <yves.junqueira at gmail.com>
# uribl Copyright 2004, Devin Carrway <qpsmtpd@devin.com>
# smtpproxy Copyright notices below
use Mail::SpamAssassin;
use Digest::MD5 qw(md5_hex);
use Net::Server::Daemonize qw(daemonize);
use Unix::Syslog qw(:macros);
use Unix::Syslog qw(:subs);
use strict;
use Getopt::Long;
use IO::File;

use lib '.';
use MSDW::SMTP::Server;
use MSDW::SMTP::Client;
use Net::DNS::Resolver;

############################################################
# Options

my %uribl_options = (
    'surblgoiano.com.br' => '',
    'multi.surbl.org'    => '',
);

# If the uribl zone is md5-enabled, the MD5 hash of the host will be 
# added to the end of this url
# my $refer_url = 'http://something/cgi-bin/grr.cgi?url=';
my $refer_url = 'http://www.surbl.org';

# leading dots whitelists sub domains
my @wl_domains_acl = qw( pop.com.br .microsoft.com hotmail.com kit.net .w3.org
  .lynx.com.br .govesa.com.br med.br);
my @wl_recipients_acl = qw ( postmaster@ abuse@ spambox@);
my $action            = "deny";
my $verbose_level     = 4;
my $check_timeout     = 30;
my $child_timeout     = 3600;
my $mode              = "transparent";
my $deny_code         = "554";
my $deny_msg          = "Content banned - Conteudo banido";
my $daemonize         = 0;
my $dns_timeout       = 5;

# daemon configs.
my $user    = "nobody";
my $group   = "nobody";
my $pidfile = "/var/run/suriproxy.pid";
############################################################
#todo
my $version = '0.11';
my %sockets;
my %checked;
my $tlds =
'zw|zm|za|yu|yt|ye|ws|wf|vu|vn|vi|vg|ve|vc|va|uz|uy|us|um|uk|ug|ua|tz|tw|tv|tt|tr|tp|to|tn|tm|tk|tj|th|tg|tf|td|tc|sz|sy|sv|su|st|sr|so|sn|sm|sl|sk|sj|si|sh|sg|se|sd|sc|sb|sa|rw|ru|ro|re|qa|py|pw|pt|ps|pro|pr|pn|pm|pl|pk|ph|pg|pf|pe|pa|org|om|nz|nu|nr|np|no|nl|ni|ng|nf|net|ne|nc|name|na|mz|my|mx|mw|mv|museum|mu|mt|ms|mr|mq|mp|mo|mn|mm|ml|mk|mil|mh|mg|md|mc|ma|ly|lv|lu|lt|ls|lr|lk|li|lc|lb|la|kz|ky|kw|kr|kn|km|ki|kh|kg|ke|jp|jo|jm|je|it|is|ir|iq|io|int|info|in|im|il|ie|id|hu|ht|hr|hn|hm|hk|gy|gw|gu|gt|gs|gr|gq|gp|gov|gn|gm|gl|gi|gh|gg|gf|ge|gd|gb|ga|fr|fo|fm|fk|fj|fi|et|es|er|eg|ee|edu|ec|dz|do|dm|dk|dj|de|cz|cy|cx|cv|cu|cr|coop|com|co|cn|cm|cl|ck|ci|ch|cg|cf|cd|cc|ca|bz|by|bw|bv|bt|bs|br|bo|bn|bm|bj|biz|bi|bh|bg|bf|be|bd|bb|ba|az|aw|au|at|as|arpa|ar|aq|ao|an|am|al|ai|ag|af|aero|ae|ad|ac';

# reversed so that .com matches before .co
my $refer_msg;
my $server;
my $res = new Net::DNS::Resolver;
$res->udp_timeout($dns_timeout);

# Check SA version
my $SA3 = -1;
if ( $Mail::SpamAssassin::VERSION < 2.6 ) {

    argh(   "Minimum Mail::SpamAssassin version is 2.6. You have"
          . "$Mail::SpamAssassin::VERSION" );
}
elsif ( $Mail::SpamAssassin::VERSION < 3.0 ) {
    do_log( 1, "Found SpamAssassin < 3.0" );
    $SA3 = 0;
}
else {
    do_log( 1, "Found SpamAssassin >= 3.0" );
    $SA3 = 1;
}

my $spamtest = Mail::SpamAssassin->new(
    {
        config_text => "use_bayes          0\n"
          . "skip_rbl_checks    1\n"
          . "bayes_path         /dev/null\n",
        local_tests_only => 1
    }
  )
  or die "Can't create SpamAssassin object";

# from GetURI:
# Remove empty/unclickable anchor tags (poisoning attempts)
# Drastically shortens output :-)
sub remove_empty_anchors($) {
    my $mail = shift;

    $mail =~ s/<[\s\=]*A[\=\s]+([^>]+)>[\s\=]*<[\=\s]*\/[\=\s]*A[\=\s]*>/[]/igs;

    return $mail;
}

sub do_log {
    my ( $pri, $msg ) = @_;
    if ( $verbose_level >= $pri ) {
        openlog "suriproxy", LOG_PID, LOG_MAIL;
        syslog LOG_INFO, "%s", $msg;
        closelog;
    }
}

sub argh {
    my ($die_msg) = defined(@_) ? @_ : '--';
    do_log( 1, "FATAL ERROR: $die_msg" );
    die "FATAL ERROR: $die_msg";
}

sub suri {
    $refer_msg = '';
    undef %checked;
    if ( defined(@_) ) {
        my ($fh) = @_;

        my (
            $dev,  $ino,   $mode,  $nlink,  $uid,     $gid, $rdev,
            $size, $atime, $mtime, $ctime,, $blksize, $blocks
          )
          = $fh->stat
          or do_log( 1, "Error: Can't check mail file: $!" )
          and return 'FAIL';

        # local ( *_, $/ );
        my $from = $server->{from};
        $from =~ s/ \w+=.*//g;

        $fh->seek( 0, 0 ) or do_log(
            1, "Error: Can't rewind message file: 
$!"
          )
          and return 'FAIL';

        do_log( 2, "Checking from=$from, to=@{$server->{to}}, SIZE=$size." );
        do_log( 6, "# of rcpts: " . @{ $server->{to} } . "." );
        my @rcpts = @{ $server->{to} };
        for (@rcpts) {
            s/[<>]//g;
            if ( lookup_acl( $_, \@wl_recipients_acl, 0 ) ) {
                do_log( 3, "Accepting message: $_ is in the whitelist" );
                return 'WHITERCPT';
            }
        }

        $SIG{ALRM} = \&timed_out_check;

        my $ret = eval {
            alarm($check_timeout);

            # test loop
            #while (1) { }

            my @message;
            while (<$fh>) {

                push @message, $_;
            }
            process_msg(@message);
            my %matches;
            while ( keys %sockets ) {
                my $c = 0;
                for my $s ( keys %sockets ) {
#              do_log(2, ".. $s");
                    unless ( $sockets{$s} ) {
                        delete $sockets{$s};
                        next;
                    }
                    next unless $res->bgisready( $sockets{$s} );
                    my $packet = $res->bgread( $sockets{$s} );
                    unless ($packet) {
                        delete $sockets{$s};
                        next;
                    }
                    for my $rr ( $packet->answer ) {
                        $matches{$s} = $rr->txtdata
                          if $rr->type eq 'TXT';
                    }
                    delete $sockets{$s};
                    $c++;
                }
                sleep 0.1 if keys %sockets and !$c;
            }
            for ( keys %matches ) {
                my ( $host, $uribl ) = split /\t/, $_;
                my $note =
"BADURI: from=$from, to=@{$server->{to}} $host in $uribl ($matches{$_}). Action=$action";
                do_log( 3, "$note." );

                if ( $action eq 'deny' ) {
                    $refer_msg =
                        $uribl_options{$uribl} =~ /m/
                      ? $refer_url . $host
                      : $refer_url;
                    return ("DENY $note $refer_msg");

                    #die "nao pode morrer";

                    #close(FILE);
                }
                else { return ("WARN $note"); }

            }

            alarm(0);
            return "OK";
        };
        alarm(0);

        if ( $@ =~ /imeout/ ) {
            return "TIMEOUT";
        }
        if ($ret) { return "$ret"; }
    }
    else {

        return "EMPTY";
    }
    return "ARGH";
}

sub process_msg(@) {
    my $mail;

    if ($SA3) {
        $mail = $spamtest->parse( \@_, 0 ) or argh("235: $!");
        my $text = $spamtest->remove_spamassassin_markup($mail)
          or argh("231: $!");
        $mail->finish();    #	 or argh("211: $!");
        $text = remove_empty_anchors($text) or argh("122: $!");
        $mail = $spamtest->parse($text)     or argh(
            "ERROR 231: Could not 
parse with SA: $!"
        );
    }
    else {
        $mail = Mail::SpamAssassin::NoMailAudit->new( data => \@_ );
        my $text    = $spamtest->remove_spamassassin_markup($mail);
        my @message = split( /$/m, remove_empty_anchors($text) );

        #$mail->delete_metadata();
        $mail = Mail::SpamAssassin::NoMailAudit->new( data => \@message );
    }

    my $status = $spamtest->check($mail);
    my @uris   = Mail::SpamAssassin::PerMsgStatus::get_uri_list($status);
    my ( $display_uri, $check_uri );

    if (@uris) { do_log( 3, "INFO: Found @uris" ); }
    foreach my $uri_text (@uris) {
        $uri_text =~ s/www\.//i;
        if (
            $uri_text =~ m{
                        \w{3,16}:/+                     # protocol
                        (?:\S+@)?                       # user/pass
                        (\d{7,})                        # raw-numeric IP
                        (?::\d+)?([/?\s]|$)             # port, slash
                                                        #  or EOL
                        }gx    # raw-numeric
          )
        {
            my @octets = (
                ( ( $1 >> 24 ) & 0xff ),
                ( ( $1 >> 16 ) & 0xff ),
                ( ( $1 >> 8 ) & 0xff ),
                ( $1 & 0xff )
            );
            $display_uri = join( '.', @octets );
            $check_uri   = join( '.', reverse @octets );
        }
        elsif (
            $uri_text =~ m{
                        \w{3,16}:/+                     # protocol
                        (?:\S+@)?                       # user/pass
                        (\d+|0[xX][0-9A-Fa-f]+)\.       # IP address
                        (\d+|0[xX][0-9A-Fa-f]+)\.
                        (\d+|0[xX][0-9A-Fa-f]+)\.
                        (\d+|0[xX][0-9A-Fa-f]+)
                        }gx
          )
        {
            my @octets = ( $1, $2, $3, $4 );

            # return any octal/hex octets in the IP addr back
            # to decimal form (e.g. http://0x7f.0.0.00001)
            for ( 0 .. $#octets ) {
                $octets[$_] =~ s/^0([0-7]+)$/oct($1)/e;
                $octets[$_] =~ s/^0x([0-9a-fA-F]+)$/hex($1)/e;
            }
            $display_uri = join( '.', @octets );
            $check_uri   = join( '.', reverse @octets );
        }
        elsif (
            $uri_text =~ m{
                        \w{3,16}:/+                     # protocol
                        (?:\S+@)?                       # user/pass
                        ([\w\-.]+\.(${tlds}))[\W]*    # hostname
                        }gx
          )
        {

            $display_uri = $1;
            $check_uri   = $1;
        }
        elsif (
            $uri_text =~ m{
                        mailto:[^@]*@                     # protocol
                        ([\w\-.]+\.(${tlds}))[\W]*    # hostname
                        }gx
          )
        {

            $display_uri = $1;
            $check_uri   = $1;
        }
        elsif ( $uri_text =~ m{([\w\-.]+\.(${tlds}))[\W]*}gx ) {
            $display_uri = $1;
            $check_uri   = $1;
        }

        if ($display_uri) {
            if ( $check_uri eq $display_uri ) {    # hostname
                my @host_domains = split /\./, $check_uri;
                while ( @host_domains >= 2 ) {
                    my $subhost = join( '.', @host_domains );
#         do_log(2,"Teste3 $subhost");
                    check_dns_host($subhost);
                    shift @host_domains;

                }
            }
            else {
           do_log(2, "Teste2 $check_uri");
  check_dns_host($check_uri) }
        }
    }
}

sub check_dns_host($) {

    my ($subhost) = @_;
#    do_log( 5, "Should I check $subhost?" );

        if (   !( $checked{"$subhost"} ) &&
      (!lookup_acl( $subhost, \@wl_domains_acl, 1 ) ) 
 )
        {
    do_log( 5, "Checking address: $subhost" );
    for my $z ( keys %uribl_options ) {
        if ( $uribl_options{$z} =~ /m/ ) {
            $subhost = md5_hex($subhost);
        }
            $sockets{"$subhost\t$z"} = $res->bgsend( "$subhost.$z.", 'txt' );
            $checked{"$subhost"}++;

        }
    }

}

sub timed_out_check {
    do_log( 1, "Error: timeout while checking message\n" );
    argh("Error: Timeout");
}

# Modified version of amavisd-new's
sub split_address($) {
    my ($mailbox) = @_;
    $mailbox =~ /^ (.*?) ( \@ (?:  \[  (?: \\. | [^\]\\] )*  \]
                                 |  [^@"<>\[\]\\\s] )*
                       ) \z/xs ? ( $1, $2 ) : ( $mailbox, '' );
}

# Modified version of amavisd-new's
sub lookup_acl($$$) {
    my ( $addr, $acl_ref, $mode ) = @_;

    #mode 1 = domain only
    #mode 0 = email

    ( ref($acl_ref) eq 'ARRAY' )
      or argh("lookup_acl: arg2 must be a list ref: $acl_ref");
    return undef if !@$acl_ref;    # empty list can't match anything
    my ( $localpart, $domain ) =
      ( $mode == 0 ) ? split_address($addr) : ( undef, $addr );
    $domain    = lc($domain);
    $localpart = lc($localpart);
    local ( $1, $2 );

    # chop off leading @ and trailing dots
    $domain = $1 if $domain =~ /^\@?(.*?)\.*\z/s;
    my ($lcaddr) = $localpart . '@' . $domain;
    my ( $found, $matchingkey, $result );
    for my $e (@$acl_ref) {
        $result      = 1;
        $matchingkey = $e;
        my ($key) = $e;
        if ( $key =~ /^(!+)(.*)\z/s ) {    # starts with an exclamation mark(s)
            $key = $2;
            $result = 1 - $result if ( length($1) & 1 );    # negate if odd
        }
        if ( $key =~ /^(.*?)\@(.*)\z/s ) {    # contains '@', check full address
            $found++
              if $localpart eq lc($1)
              && ( length( lc($2) ) == 0 || $domain eq lc($2) );
        }
        elsif ( $key =~ /^\.(.*)\z/s ) {      # leading dot: domain or subdomain
            my ($key_t) = lc($1);
            $found++ if $domain eq $key_t || $domain =~ /\Q$key_t\E\z/s;
        }
        else {    # match domain (but not its subdomains)
            $found++ if $domain eq lc($key);
        }
        last if $found;
    }
    $matchingkey = $result = undef if !$found;
    if ( !defined( $checked{"$addr acl"} ) ) {
        do_log(
            ( $found ? 3 : 6 ),
            "Whitelist check: $addr"
              . (
                !$found
                ? "doesn't match"
                : " matches entry \"$matchingkey\""
              )
        );
        $checked{"$addr acl"}++;
    }
    !wantarray ? $result : ( $result, $matchingkey );
}

#
#   This code is Copyright (C) 2001 Morgan Stanley Dean Witter, and
#   is distributed according to the terms of the GNU Public License
#   as found at <URL:http://www.fsf.org/copyleft/gpl.html>.
#
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
# Written by Bennett Todd <bet@rahul.net>

my $syntax =
    "syntax: $0 [--children=16] [--minperchild=100] "
  . "[--maxperchild=200] [--debugtrace=undef] "
  . "listen.addr:port talk.addr:port\n";

my $children    = 3;
my $minperchild = 100;
my $maxperchild = 200;
my $debugtrace  = undef;
GetOptions(
    "children=n"    => \$children,
    "minperchild=n" => \$minperchild,
    "maxperchild=n" => \$maxperchild,
    "debugtrace=s"  => \$debugtrace
  )
  or die $syntax;

die $syntax unless @ARGV == 2;
my ( $srcaddr, $srcport ) = split /:/, $ARGV[0];
my ( $dstaddr, $dstport ) = split /:/, $ARGV[1];
die $syntax unless defined($srcport) and defined($dstport);

# This should allow a kill on the parent to also blow away the
# children, I hope
my %children;
use vars qw($please_die);
$please_die = 0;
$SIG{TERM} = sub { $please_die = 1; };
$server = MSDW::SMTP::Server->new( interface => $srcaddr, port => $srcport );

if ( $daemonize != 0 ) {
    daemonize( $user, $group, $pidfile )
      or argh("Error: could not detach: $!");
}
else {
    print "Suriproxy starting. Staying in the foreground (see \$daemonize).\n";
}

do_log( 1, "Suriproxy v$version starting." );

# This block is the parent daemon, never does an accept, just herds
# a pool of children who accept and service connections, and
# occasionally kill themselves off

PARENT: while (1) {
    while ( scalar( keys %children ) >= $children ) {
        my $child = wait;
        delete $children{$child} if exists $children{$child};
        if ($please_die) { kill 15, keys %children; exit 0; }
    }
    my $pid = fork;
    argh("$0: fork failed: $!\n") unless defined $pid;
    last PARENT if $pid == 0;
    $children{$pid} = 1;
    select( undef, undef, undef, 0.1 );
    if ($please_die) { kill 15, keys %children; exit 0; }
}

# This block is a child service daemon. It inherited the bound
# socket created by SMTP::Server->new, it will service a random
# number of connection requests in [minperchild..maxperchild] then
# exit

my $lives = $minperchild + ( rand( $maxperchild - $minperchild ) );
my %opts;

#suri settings

$res->udp_timeout(5);

if ( defined $debugtrace ) {
    $opts{debug} = IO::File->new(">$debugtrace");
    $opts{debug}->autoflush(1);
}

while (1) {
    $server->accept(%opts);

    local $SIG{ALRM} = sub { die "Child server process timed out!\n" };

    alarm($child_timeout);

    my $client =
      MSDW::SMTP::Client->new( interface => $dstaddr, port => $dstport );
    my $banner = $client->hear
      or argh("FATAL ERROR: 1202 $!");

    #    $banner = "220 $debugtrace" if defined $debugtrace;
    $server->ok($banner)
      or argh("FATAL ERROR: 1222 $!");

    while ( my $what = $server->chat ) {

        if ( $what eq '.' ) {
            my $check = &suri( $server->{data} );

            $server->{data}->seek( 0, 0 )
              or argh("FATAL ERROR: 1142 $!");

            do_log( 5, "Check result: $check" );
            if ( $check =~ /^DENY/ ) {
                if ( $mode eq 'transparent' ) {
                    $server->ok("$deny_code $deny_msg $refer_msg.")
                      or argh("FATAL ERROR: 1132 $!");

                    last;

                    #next; #todo: continue the dialog
                }
                else {
                    $server->ok("250 OK - Discarded: $check.")
                      or argh("FATAL ERROR: 1102 $!");

                    #        last;
                    next;
                }
            }
            else {
                $client->yammer( $server->{data} )
                  or argh("FATAL ERROR: 1103 $!");
            }

        }
        else {
            $client->say($what)
              or argh("FATAL ERROR: 1104 $!");

        }

        #from spampd
        my $destresp = $client->hear
          or argh("FATAL ERROR: 1122 $!");
        $server->ok($destresp)
          or argh("Error in server->ok(client->hear): $!");

        if ( $server->{state} =~ /^data/i and $destresp =~ /^[45]\d{2}/ ) {
            $server->{state} = "err_after_data";
        }
        alarm($child_timeout);
    }
    $client = undef;
    delete $server->{"s"};
    exit 0 if $lives-- <= 0;
}

